### PR TITLE
convert.py: Hot Fix with VocabFactory Integration

### PR DESCRIPTION
**Pull request by teleprint-me**
#### Overview
This pull request introduces significant improvements to the `convert.py` script in the `convert-py` branch of the `llama.cpp` project. Key changes include the introduction of the `VocabFactory` class for streamlined vocabulary handling, refactoring of the `OutputFile` class for better output management, and various user interface improvements.

#### Key Changes

1. **Added BpeVocab and SentencePieceVocab**: Reintroduced the original `BpeVocab` and `SentencePieceVocab` classes to ensure backward compatibility and support for the original llama models. This addition ensures that existing workflows remain functional and uninterrupted.

2. **New HfVocab Supporting AutoTokenizer Vocab**: Replaced the previous `VocabLoader` with the new `HfVocab` class. This class streamlines the integration of `transformers.AutoTokenizer`, ensuring forward compatibility and more efficient handling of Hugging Face tokenizers.

3. **VocabFactory Implementation**: Implemented a new `VocabFactory` class to encapsulate the logic for handling different types of vocabularies. This enhancement significantly improves the flexibility and maintainability of vocabulary loading and processing, accommodating a wider range of model types and formats.

4. **Argument Parser and CLI Improvements**: The command-line interface has been thoroughly reorganized and refined. Improvements include better argument naming, enhanced error handling, and a more intuitive user interface, all of which contribute to improved usability and a smoother user experience.

5. **Code Quality and Readability**: Dedicated efforts were made to enhance code readability, organization, and documentation throughout the `convert.py` script. These improvements align with best practices in Python development, aiming to facilitate easier maintenance and future enhancements.

The original core llama models should now convert as expected with this hot fix.

[Screencast from 2024-01-07 15-30-21.webm](https://github.com/ggerganov/llama.cpp/assets/77757836/da955db6-3f26-43c6-867b-199f3e988352)

#### Migration Path for `transformers.AutoTokenizer` Support
To further streamline the functionality of `convert.py` and align with the modular design approach, I recommend migrating the support for `transformers.AutoTokenizer` from `convert.py` to a dedicated script, `convert-hf-to-gguf.py`. This migration will allow for more focused handling of Hugging Face models and tokenizers, ensuring each script maintains a clear and distinct purpose. This separation of concerns will not only simplify the maintenance but also enhance the scalability of our tooling for different model formats.

#### Request for Review
I kindly request a review of these changes and welcome any feedback or suggestions. The improvements aim to make the `convert.py` script more robust, flexible, and user-friendly, and I believe they will significantly benefit the `llama.cpp` project.

### CODE REVIEW

File: convert.py [hunk 1]
```diff
@@ -1178,32 +1602,111 @@ def do_dump_model(model_plus: ModelPlus) -> None:
...
+    )
+
+    parser.add_argument(
+        "--vocab-only",
+        action="store_true",
+        help="Extract and output only the vocabulary",
+    )
+
+    parser.add_argument(
+        "--out-type",
```
Comment (ggerganov) [inline convert.py:1652]: Update `--outtype` instances in the main README.md with the new `--out-type`

File: convert.py [hunk 2]
```diff
@@ -841,30 +1099,42 @@ def bounded_parallel_map(func: Callable[[In], Out], iterable: Iterable[In], conc
 def check_vocab_size(params: Params, vocab: Vocab, pad_vocab: bool = False) -> None:
     if params.n_vocab != vocab.vocab_size:
         if params.n_vocab == vocab.vocab_size:
```
Comment (ggerganov) [inline convert.py:1101]: This is not related to the changes in this PR, but the second `if` is never true

File: convert.py [hunk 3]
```diff
@@ -291,129 +362,254 @@ def loadOriginalParamsJson(model: LazyModel, config_path: Path) -> Params:
...
+            n_experts=n_experts,
+            n_experts_used=n_experts_used,
+            f_norm_eps=config["norm_eps"],
+            f_rope_freq_base=config.get("rope_theta", f_rope_freq_base),
         )
 
     @staticmethod
-    def load(model_plus: ModelPlus) -> Params:
-        hf_config_path   = model_plus.paths[0].parent / "config.json"
+    def load(model_plus: ModelPlus) -> "Params":
```
Comment (ggerganov) [inline convert.py:384]: Is this correctly changed from `Params` to `"Params"`?

**Review by ggerganov — APPROVED** [review]
The vanilla LLaMA v1 and v2 models that I have downloaded now work after updating the `n_vocab` in the `params.json` to 32000 (the original file has -1 for some reason and we generate a warning) so this is good

Btw, in this case it would have been better to first submit a minimal PR that fixes #4493 and a follow-up PR with code-style changes, because it is more difficult to review now

**Comment by teleprint-me** [comment]
That's why I formatted it last. It would be easier to rollback if it became an issue. 

I intentionally did individual commits, rather than one large commit, as I usually do, so it would be easier to track the changes and rollback any commits if needed.

Will keep in mind for future. Would definitely recommend a formatter though. Ruff can be customized. 

I tested it with llama, mistral, mixtral, and deepseek. Regardless, glad it's working again.

**Comment by ggerganov** [comment]
Ok thanks. After fixing the README and the flake CI, we can merge

**Comment by teleprint-me** [comment]
@ggerganov

Given the uncertainties around the `n_vocab` parameter being set to `-1`, I've taken a cautious approach in this PR to ensure immediate stability. This issue will need further investigation post-release to understand its full impact. It's important to note that the conversion process appears to succeed with `vocab_size` set to either `-1` or `32000`, but the implications of this success are currently unclear.

File: convert.py [hunk 4]
```diff
@@ -153,65 +183,85 @@ def type_for_tensor(self, name: str, tensor: LazyTensor) -> DataType:
...
+    f_rope_freq_base: Optional[float] = None
+    f_rope_scale: Optional[float] = None
+    n_orig_ctx: Optional[int] = None
+    rope_finetuned: Optional[bool] = None
+
+    ftype: Optional[GGMLFileType] = None
 
     # path to the directory containing the model files
-    path_model: Path | None = None
+    path_model: Optional[Path] = None
```
Comment (cebtenzzre) [inline convert.py:206]: Why aren't we using [PEP 604](https://peps.python.org/pep-0604/) types anymore?

File: convert.py (hunk 3 above)
Comment (cebtenzzre) [reply convert.py:365]: This is incorrect - we are using `from __future__ import annotations` above, so type annotations have no reason to be quoted.

File: convert.py [hunk 5]
```diff
@@ -1145,30 +1355,95 @@ def load_some_model(path: Path) -> ModelPlus:
...
+            file_key = "tokenizer.model" if vocabtype == "spm" else "vocab.json"
+            if self.files[file_key]:
+                return self.files[file_key]
+            else:
+                raise FileNotFoundError(f"{vocabtype} {file_key} not found.")
+        elif vocabtype == "hfft":
+            # For Hugging Face Fast Tokenizer, return the directory path instead of a specific file
+            return self.path
+        else:
+            raise ValueError(f"Unsupported vocabulary type {vocabtype}")
```
Comment (cebtenzzre) [inline convert.py:1389]: Several examples of else-after-return here, which is an anti-pattern.

**Comment by cebtenzzre** [comment]
How do we agree as a community on a code style for the Python scripts? I don't want this to become a tug-of-war. I have so far tried to keep my changes minimal (and tried not to be opinionated about whitespace), but this PR reverts many of them for no clear reason.

**Comment by teleprint-me** [comment]
@cebtenzzre, I appreciate your point about maintaining a consistent code style across the project. I agree that the best way forward is to decide as a community on a linter and formatter for the Python scripts. This approach would help us avoid any style conflicts and ensure uniformity in the codebase.

My use of PEP-0484 guidelines in this PR was based on habit and not intended to override or challenge previous stylistic choices. I apologize if it seemed to revert any changes unintentionally. 

I'm fully open to adapting to the community's agreed-upon style standards. To facilitate this, I suggest we could include configuration files for the chosen linter and formatter in our repository. This would automate the enforcement of our style rules and make it easier for all contributors to adhere to them consistently.

We can work together to establish any standards and integrate them into our workflow. I'm looking forward to your thoughts and suggestions on this.

**Comment by ggerganov** [comment]
I guess now is a good time to apply an auto-formatting tool to the python code as I believe there aren't any major developments going on, so conflicts should be minimal. I don't have preference so open to suggestions

@cebtenzzre The main reason to merge this PR is because the `convert.py` script has been broken for a long time and I wanted to get it fixed. I would've preferred a much smaller diff, but fixing the conversion seemed more important

File: convert.py (hunk 3 above)
Comment (teleprint-me) [reply convert.py:365]: @cebtenzzre, I understand your point about the `from __future__ import annotations` directive and how it affects the need for string-based annotations. The use of `"Params"` as a forward reference was intended as a precaution against potential circular dependencies, as outlined in PEP 563 and PEP 649.

However, I respect the established practices of this project. If the consensus is that explicit string annotations are not necessary or preferred in the codebase, I'm willing to adapt to that standard. I'll always appreciate your guidance and expertise on any matter.

**Comment by teleprint-me** [comment]
`ruff` and `black` are both solid choices for auto-formatting. `ruff` has been gaining popularity recently, while `black` is more established and widely used. Since `flake8` is already part of the linter setup, it might be good to align with that, especially if it's integrated into the CI pipeline.

What do you think @cebtenzzre ?

**Pull request merged**